qrexec/policy/parser.py

```diff
@@ -523,7 +525,8 @@ def from_ask_resolution(cls, ask_resolution, *, target):
             if dispvm:
                 self.cleanup_dispvm(target)
         if process.returncode != 0:
-            raise AccessDenied('qrexec-client failed')
+            # Return "request refused" to the remote end.
+            raise AccessDenied('qrexec-client failed', notify=False)
```

Comment:
Shouldn't this be `notify=True` (as with other denied cases)?

Reply:
This is a confusing corner case, because this happens after we showed an "Allowed" notification, so the user might see both. Do you think that's better than failing silently?

Reply:
I thought about this and I actually think we should not notify here.

1. `qrexec-client` can also return a non-zero exit code in case the call went through, but the program on remote end exited with non-zero status (though not always: in some cases `qrexec-client` returns early). So, while we could add a special notification ("qubes.Service failed"), it would be pretty confusing, because it wouldn't be shown in every case.

2. This is pure speculation but source program might expect a non-zero exit code as normal part of operation, so it's not even a failure (imagine "qvm-run grep ..."). In any case, I think it's better to leave handling the failure to the source.

3. The situation this piece of code is actually supposed to handle is when something goes wrong with setting up the connection. I think ensuring that the source gets `MSG_CONNECTION_REFUSED` is right, but treating this situation as a "acces denied" isn't.

So, for now I would use another exception here (`raise ConnectionFailed`?) to make it clear that this is not a policy refusal. And leave a comment explaining that this is a workaround and we don't currently distinguish cases 1/2 from case 3.

Reply:
This is indeed confusing corner case. The issue with just `MSG_CONNECTION_REFUSED` and no notification is that the source side may expect the notification was already shown for `MSG_CONNECTION_REFUSED`, so doing own error notification would be wrong in most cases.

Maybe better solution would be checking if target's qrexec-daemon is alive at the final policy evaluation step (when we know the actual target already), before answering "allow" at all? It may not handle all the failure modes, but the most common will and will also avoid confusing qrexec-client exit code cases.
In the future (as discussed somewhere already), we should move VM<->VM connection negotiation to the qrexec-daemon itself, so qrexec-client's only responsibility will be connecting local process (or its own stdin/stdout) with vchan and the confusing exit code case should be gone.

Reply:
Checking if the daemon is alive would mean trying to connect to it, right? Because the daemon process might be even alive but in a "reconnecting" state where it doesn't accept connections.

With how the code is laid out currently, the current connection happens too late. We could connect early and pass the FD to qrexec-client. Alternatively, maybe add a "ping the daemon" step to the policy-daemon, that just opens a connection to see if it would be accepted? (To be honest both solutions seem kind of ugly).

Reply:
What about reserving yet another exit code (125?) in qrexec-client for "connection failed" and raise exception only in that case?

Reply:
I like it. I guess it might collide with service's exit code, but it sounds unlikely.

I would still display a special notification ("resolution": "connection_failed", not "resolution": "deny"; and another message for the user).

Reply:
Yes, different notification sounds good.

Reply:
Okay, no, I don't like it anymore :( There are too many `exit()`s in qrexec-client and I really don't want to add another "feature" to it right now.

Reply:
Oh, but the other way around is easier - I can add an option to exit with 0 even if child/remote exited with non-zero.